File: clients/go/decode.go
package oresapidocs

import (
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"
	"unicode/utf8"
)

func DecodeCall(payload []byte) (Call, error) {
	raw, err := decodeObject(payload, callFields)
	if err != nil {
		return Call{}, err
	}
	var call Call
	if err = required(raw, "v", &call.Version); err != nil {
		return Call{}, err
	}
	var op string
	if err = required(raw, "op", &op); err != nil || op != "call" {
		return Call{}, errors.New("expected op call")
	}
	if err = required(raw, "id", &call.ID); err != nil {
		return Call{}, err
	}
	if err = required(raw, "key", &call.Key); err != nil {
		return Call{}, err
	}
	if err = optionalPtr(raw, "transport", &call.Transport); err != nil {
		return Call{}, err
	}
	call.Path = rawJSON(raw, "path")
	call.Query = rawJSON(raw, "query")
	call.Body = rawJSON(raw, "body")
	if err = optionalPtr(raw, "traceId", &call.TraceID); err != nil {
		return Call{}, err
	}
	if err = optionalPtr(raw, "spanId", &call.SpanID); err != nil {
		return Call{}, err
	}
	return call, call.Validate()
}
func DecodeReceipt(payload []byte) (Receipt, error) {
	raw, err := decodeObject(payload, receiptFields)
	if err != nil {
		return Receipt{}, err
	}
	var receipt Receipt
	if err = required(raw, "v", &receipt.Version); err != nil {
		return Receipt{}, err
	}
	var op string
	if err = required(raw, "op", &op); err != nil || op != "receipt" {
		return Receipt{}, errors.New("expected op receipt")
	}
	if err = required(raw, "id", &receipt.ID); err != nil {
		return Receipt{}, err
	}
	if err = required(raw, "key", &receipt.Key); err != nil {
		return Receipt{}, err
	}
	if err = optionalPtr(raw, "transport", &receipt.Transport); err != nil {
		return Receipt{}, err
	}
	if err = required(raw, "ok", &receipt.OK); err != nil {
		return Receipt{}, err
	}
	if err = optionalPtr(raw, "status", &receipt.Status); err != nil {
		return Receipt{}, err
	}
	receipt.Body = rawJSON(raw, "body")
	receipt.Error = rawJSON(raw, "error")
	if err = optionalPtr(raw, "traceId", &receipt.TraceID); err != nil {
		return Receipt{}, err
	}
	if err = optionalPtr(raw, "spanId", &receipt.SpanID); err != nil {
		return Receipt{}, err
	}
	return receipt, receipt.Validate()
}
func decodeObject(payload []byte, allowed map[string]struct{}) (map[string]json.RawMessage, error) {
	if len(payload) > MaxFrameBytes {
		return nil, fmt.Errorf("frame is %d bytes, over the %d limit", len(payload), MaxFrameBytes)
	}
	if !utf8.Valid(payload) {
		return nil, errors.New("frame is not UTF-8")
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, fmt.Errorf("frame is not JSON: %w", err)
	}
	if raw == nil {
		return nil, errors.New("frame must be a JSON object")
	}
	var unknown []string
	for key := range raw {
		if _, ok := allowed[key]; !ok {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		sort.Strings(unknown)
		return nil, fmt.Errorf("unknown frame member(s): %s", strings.Join(unknown, ", "))
	}
	return raw, nil
}
func required(raw map[string]json.RawMessage, name string, target any) error {
	value, ok := raw[name]
	if !ok {
		return fmt.Errorf("missing frame member %s", name)
	}
	if err := json.Unmarshal(value, target); err != nil {
		return fmt.Errorf("%s has the wrong type", name)
	}
	return nil
}
func optionalPtr[T any](raw map[string]json.RawMessage, name string, target **T) error {
	value, ok := raw[name]
	if !ok {
		return nil
	}
	var parsed T
	if err := json.Unmarshal(value, &parsed); err != nil {
		return fmt.Errorf("%s has the wrong type", name)
	}
	*target = &parsed
	return nil
}
func rawJSON(raw map[string]json.RawMessage, name string) OptionalJSON {
	value, ok := raw[name]
	return OptionalJSON{Present: ok, Value: clone(value)}
}

File: rust/src/rpc_v1/decode.rs
pub fn decode_rpc_v1_call(payload: &[u8]) -> Result<RpcV1Call, SchemaError> {
    let value = decode_object(payload, "rpc-call", &CALL_FIELDS)?;
    validate_rpc_call(&Value::Object(value.clone()))?;
    let version = required_u64(&value, "v", "rpc-call")?;
    if version != u64::from(RPC_V1_VERSION) {
        return instance("rpc-call", format!("unsupported RPC version {version}"));
    }
    if required_string(&value, "op", "rpc-call")? != "call" {
        return instance("rpc-call", "expected op call");
    }
    let call = RpcV1Call {
        id: required_string(&value, "id", "rpc-call")?.to_owned(),
        key: required_string(&value, "key", "rpc-call")?.to_owned(),
        transport: optional_transport(&value, "rpc-call")?,
        path: optional_object(&value, "path", "rpc-call")?,
        query: optional_object(&value, "query", "rpc-call")?,
        headers: optional_object(&value, "headers", "rpc-call")?,
        body: value
            .get("body")
            .cloned()
            .map_or_else(OptionalJson::absent, OptionalJson::present),
        trace_id: optional_string(&value, "traceId", "rpc-call")?,
        span_id: optional_string(&value, "spanId", "rpc-call")?,
    };
    call.validate()?;
    Ok(call)
}

pub fn decode_rpc_v1_receipt(payload: &[u8]) -> Result<RpcV1Receipt, SchemaError> {
    let value = decode_object(payload, "rpc-receipt", &RECEIPT_FIELDS)?;
    validate_rpc_receipt(&Value::Object(value.clone()))?;
    let version = required_u64(&value, "v", "rpc-receipt")?;
    if version != u64::from(RPC_V1_VERSION) {
        return instance("rpc-receipt", format!("unsupported RPC version {version}"));
    }
    if required_string(&value, "op", "rpc-receipt")? != "receipt" {
        return instance("rpc-receipt", "expected op receipt");
    }
    let status = match value.get("status") {
        None => None,
        Some(raw) => Some(
            raw.as_u64()
                .and_then(|value| u16::try_from(value).ok())
                .ok_or_else(|| schema_error("rpc-receipt", "status has the wrong type"))?,
        ),
    };
    let receipt = RpcV1Receipt {
        id: required_string(&value, "id", "rpc-receipt")?.to_owned(),
        key: required_string(&value, "key", "rpc-receipt")?.to_owned(),
        transport: optional_transport(&value, "rpc-receipt")?,
        ok: value
            .get("ok")
            .and_then(Value::as_bool)
            .ok_or_else(|| schema_error("rpc-receipt", "ok has the wrong type"))?,
        status,
        body: value
            .get("body")
            .cloned()
            .map_or_else(OptionalJson::absent, OptionalJson::present),
        error: optional_object(&value, "error", "rpc-receipt")?,
        trace_id: optional_string(&value, "traceId", "rpc-receipt")?,
        span_id: optional_string(&value, "spanId", "rpc-receipt")?,
    };
    receipt.validate()?;
    Ok(receipt)
}

pub fn rpc_v1_call_from_ndjson(payload: &[u8]) -> Result<RpcV1Call, SchemaError> {
    decode_rpc_v1_call(single_ndjson(payload, "rpc-call")?)
}

pub fn rpc_v1_receipt_from_ndjson(payload: &[u8]) -> Result<RpcV1Receipt, SchemaError> {
    decode_rpc_v1_receipt(single_ndjson(payload, "rpc-receipt")?)
}

pub fn assert_rpc_v1_receipt_for_call(
    call: &RpcV1Call,
    receipt: &RpcV1Receipt,
) -> Result<(), SchemaError> {
    call.validate()?;
    receipt.validate()?;
    if receipt.id != call.id {
        return instance("rpc-receipt", "receipt id does not match call id");
    }
    if receipt.key != call.key {
        return instance("rpc-receipt", "receipt key does not match call key");
    }
    if let (Some(left), Some(right)) = (call.transport, receipt.transport) {
        if left != right {
            return instance(
                "rpc-receipt",
                "receipt transport does not match call transport",
            );
        }
    }
    Ok(())
}

pub fn split_rpc_v1_length_prefixed(buffer: &[u8]) -> Result<(Vec<&[u8]>, &[u8]), SchemaError> {
    split_length_prefixed(buffer)
}

#[derive(Debug, Default)]
pub struct RpcV1Correlator {
    prefix: String,
    next: u64,
}

impl RpcV1Correlator {
    pub fn new(prefix: impl Into<String>) -> Result<Self, SchemaError> {
        let prefix = prefix.into();
        if prefix.chars().count() >= 128 {
            return instance(
                "rpc-call",
                "correlation prefix must contain fewer than 128 Unicode scalars",
            );
        }
        Ok(Self { prefix, next: 0 })
    }

    pub fn take(&mut self) -> Result<String, SchemaError> {
        self.next = self
            .next
            .checked_add(1)
            .ok_or_else(|| schema_error("rpc-call", "correlation id counter exhausted"))?;
        let value = format!("{}{}", self.prefix, self.next);
        validate_string(&value, "correlation id", 128, "rpc-call")?;
        Ok(value)
    }
}
